src/java/org/apache/cassandra/config/GuardrailsOptions.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.config;
+
+import java.util.Set;
+import java.util.function.Consumer;
+
+import org.apache.cassandra.db.guardrails.Guardrails;
+import org.apache.cassandra.db.guardrails.GuardrailsConfig;
+
+/**
+ * Configuration settings for guardrails populated from the Yaml file.
+ *
+ * <p>Note that the settings here must only be used to builf the {@link GuardrailsConfig} class and not directly by the
+ * code checking each guarded constraint. That code should use the higher level abstractions defined in
+ * {@link Guardrails}).
+ *
+ * <p>This contains a main setting, {@code enabled}, controlling if guardrails are globally active or not, and
+ * individual settings to control each guardrail.
+ *
+ * <p>We have 2 variants of guardrails, soft (warn) and hard (fail) limits, each guardrail having either one of the
+ * variants or both. Note in particular that hard limits only make sense for guardrails triggering during query
+ * execution. For other guardrails, say one triggering during compaction, failing does not make sense.
+ *
+ * <p>Additionally, each individual setting should have a specific value (typically -1 for numeric settings),
+ * that allows to disable the corresponding guardrail.
+ *
+ * <p>For consistency, guardrails based on a simple numeric threshold should use the naming scheme
+ * {@code <what_is_guarded>_warn_threshold} for soft limits and {@code <what_is_guarded>_fail_threshold} for hard
+ * ones, and if the value has a unit, that unit should be added at the end (for instance,
+ * {@code <what_is_guarded>_fail_threshold_in_kb}). For "boolean" guardrails that disable a feature, use
+ * {@code <what_is_guarded_enabled}. Other type of guardrails can use appropriate suffixes but should start with
+ * {@code <what_is_guarded>}.
+ */
+public class GuardrailsOptions
```

Comment:
it is not you thing, I feel it the same way. Invalid request is too ... harsh. Why cant we throw GuardrailsException? Because drivers are not prepared for it or what?

I like nested configs but only in case if there is majority of cases when they will not consist of 1 treshold only.

Reply:
> Why cant we throw GuardrailsException? Because drivers are not prepared for it or what?

Sadly yes.  We can throw w/e we want, but the server serialization logic can't handle it...  On coordinator the only things that can be thrown are things which `org.apache.cassandra.transport.Message.Codec#encode` is able to handle.    Also sadly client doesn't allow us to add to `org.apache.cassandra.exceptions.ExceptionCode`, so can't add new exceptions without client supporting (and assuming client supports, getting users on that version... 😢 ); client rejects any code it doesn't know explicitly...

There was a bug I had to fix with exceptions being thrown in coordinator... we can NOT throw w/e we want, we can only throw w/e `org.apache.cassandra.transport.Message.Codec#encode` can understand... if you don't then client gets the more fun "something blew up... blame everyone but yourself"... its why the track warning tests all double check client integration -_-.

Reply:
> I like nested configs but only in case if there is majority of cases when they will not consist of 1 treshold only.

yeah, I agree that this looks dumb

```
table:
  threshold: 10
```

should just be `table_threshold: 10`

but given the warn/fail nature of this, I do prefer

```
guardrails:
  enabled: true
  tables:
    warn_threshold: 100
    fail_threshold: 9001
  local_read_size:
    warn_threshold: 10kb
    fail_threshold: 1mb
```

Reply:
It's hard to tell in advance but I suspect that there will be a good number, even a majority of guardrails using only one of the limits. For starters, the guardrails that enable/disable a flag only have one single property to nest. Also, guardrails that are not associated to a user query will probably only have the soft limit. Conversely, guardrails such as, for example, the max value of a column will only have the hard limit because it probably doesn't make sense to warn users every time they write a column with a size that is close to the hard limit. 

Taking some examples from the CEP, this configuration:
```
guardrails:
  enabled: false
  column_value_size_fail_threshold_in_kb: -1
  columns_per_table_fail_threshold: -1
  secondary_index_per_table_fail_threshold: -1
  materialized_view_per_table_fail_threshold: -1
  tables_warn_threshold: -1
  tables_fail_threshold: -1
  table_properties_disallowed:
  table_properties_ignored:
  write_consistency_levels_disallowed:
  partition_size_warn_threshold_in_mb: -1
  partition_keys_in_select_fail_threshold: -1
  disk_usage_percentage_warn_threshold: -1
  disk_usage_percentage_fail_threshold: -1
  in_select_cartesian_product_fail_threshold: -1
  user_timestamps_enabled: true
  read_before_write_list_operations_enabled: true
```
Would look like this:
```
guardrails:
  enabled: false
  column_value_size:
    fail_threshold_in_kb: -1
  columns_per_table:
    fail_threshold: -1
  secondary_index_per_table:
    fail_threshold: -1
  materialized_view_per_table:
    fail_threshold: -1
  tables:
    warn_threshold: -1
    fail_threshold: -1
  table_properties:
    disallowed:
    ignored:
  write_consistency_levels:
    disallowed:
  partition_size:
    warn_threshold_in_mb: -1
  partition_keys_in_select:
    fail_threshold: -1
  disk_usage_percentage:
    warn_threshold: -1
    fail_threshold: -1
  in_select_cartesian_product:
    fail_threshold: -1
  user_timestamps:
    enabled: true
  read_before_write_list_operations:
    enabled: true
```
There is also the question of the consistency with JMX and virtual tables. Probably we don't want to have the nested structure in JMX and VTs, I guess we'll prefer plain names for those. For example, track_warnings have nested configs in the yaml, but the associated JMX methods still have plain names such as `setCoordinatorLargeReadWarnThresholdKB` and `setCoordinatorLargeReadAbortThresholdKB`. I guess that for virtual tables we would also choose similar plain names, unless we want to suffer with collections/UDTs. 

I think it would be just easier to understand if we always use not-nested names, but I'm open to change the yaml format if we agree that nested names are preferable.

Reply:
> also, wonder fail vs abort? Right now client protocol doesn't really let us do failures which tells the user "this is your issue", so the client always sees "Cassandra bug... " then you check the logs and might see something saying "you did this...". I personally like abort to be clear that this isn't our issue, its where we give up and yell at people.

Making it clear that we are cancelling makes sense to me. I gave it a go in [this separate commit](https://github.com/adelapena/cassandra/commit/e2d2186f053f2befe3181a05abab55436061cc81) to see how it looks. Not sure whether we should use the term `abort` instead of `fail` only in the config properties and user-facing messages, or apply it everywhere (`fail` and `failure` are used everywhere).

Reply:
based on your example, this is the case when there is more 1 item configs rather than having majority of them as I suggested above so I would just leave it like it is now - flat.

Reply:
> the guardrails that enable/disable a flag only have one single property to nest.

Yep, stuff like enabled I wouldn't nest, so in your example `user_timestamps_enabled: true` (enabled as the prefix/postfix doesn't matter to me, just the example); I am a fan of only nesting when multiple configs tie to a single thing.

> Also, guardrails that are not associated to a user query will probably only have the soft limit. 

what case is this?  As I said just above I am cool for not nesting when there is only a single config.

> Conversely, guardrails such as, for example, the max value of a column will only have the hard limit because it probably doesn't make sense to warn users every time they write a column with a size that is close to the hard limit.

I disagree here as users can't act until they know, and if we don't tell them early they can't fix things.  I have seen cases where users didn't know that their data model caused issues, or an upstream thing had a bug now their columns started to grow; if we don't warn till it's too late we will trigger a user facing outage.  This gets worse when multiple applications share the same data, as now we have a more complex outage and 0 warning.

Based off that, this is how I see the configs

```
guardrails:
  enabled: false
  user_timestamps_enabled: true
  read_before_write_list_operations_enabled: true
  write_consistency_levels_disallowed: []
  column_value_size:
    warn_threshold_in_kb: -1
    fail_threshold_in_kb: -1
  columns_per_table:
    warn_threshold: -1
    fail_threshold: -1
  secondary_index_per_table:
    warn_threshold: -1
    fail_threshold: -1
  materialized_view_per_table:
    warn_threshold: -1
    fail_threshold: -1
  tables:
    warn_threshold: -1
    fail_threshold: -1
  table_properties:
    disallowed: []
    ignored: []
  partition_size:
    warn_threshold_in_mb: -1
    fail_threshold_in_mb: -1
  partition_keys_in_select:
    warn_threshold: -1
    fail_threshold: -1
  disk_usage_percentage:
    warn_threshold: -1
    # even though this is likely not tied to a query, being able to limit queries from writing once the disk is "too full" can be valuable
    fail_threshold: -1
  in_select_cartesian_product:
    warn_threshold: -1
    fail_threshold: -1
```

for me, I see far more nested than I do singular (very small number where I can see arguing singular).

Reply:
> > Also, guardrails that are not associated to a user query will probably only have the soft limit.

> what case is this? As I said just above I am cool for not nesting when there is only a single config.

For example the one mentioned before, a guardrail that warns on partition size during compaction. We don't want to abort compaction in that case, just warn.

Reply:
I am actively working on a patch to warn/abort queries if they touch large partitions =)

Reply:
> I am actively working on a patch to warn/abort queries if they touch large partitions =)
Do it with guardrails! :)

Independently of how that warn/abort is done, I guess we can agree that there could be guardrails that happen on compaction/repair/etc. that are not associated to user requests, can we? If that's the case, those guardrails shouldn't have an abort threshold. How would the partially nested configuration look like for those vales? Maybe this way:
```yaml
guardrails:
  enabled: false
  column_value_size:
    warn_threshold_in_kb: -1
    fail_threshold_in_kb: -1
  partition_size_in_mb: -1
```
Or this other way:
```yaml
guardrails:
  enabled: false
  column_value_size:
    warn_threshold_in_kb: -1
    fail_threshold_in_kb: -1
  partition_size:
    warn_threshold_in_mb: -1
    # no fail threshold for partitions
```

Reply:
> Do it with guardrails! :)

CASSANDRA-16310 is the prereq to doing it; once that is merged can send the PR.  Currently it is in track_warnings, so once guardrails can handle the track_warnings case then can migrate; this sadly has to be in replica as coordinator may not handle the partition being queried =/.

> I guess we can agree that there could be guardrails that happen on compaction/repair/etc. that are not associated to user requests, can we?

Let's frame this question differently, for cases like `partition_size` which may _over time_ have a user request integration hook added, _should_ the warn thresholds match on user request and repair/compaction? I agree fail for repair/compaction shouldn't happen.

The reason I frame this question this way is that we are talking about a user/operator facing API, so if we leak implementation details into the API we have to live with that for years.

Can I see cases where repair/compaction is the only thing which will ever have the guardrail?  Sure, but this should be clear in the name, maybe something like the following

```
guardrails:
  compaction: # repair validation runs in compaction... but can see that as a confusing thing...
    partition_size:
      warn_threshold: 1gb
```

If we add a guard rail and don't "scope" it to a subsystem, then it is safe to assume that a user will think it's global to all subsystems; so cases like `partition_size` should logically impact read/write requests.  For these non-scoped cases, we may not offer fail day 1, but we may add it later on, so have to worry about how we evolve it.  So for that example, if we only allowed the following day 1, we are later able to "extend" it to offer fail when we are ready to add it to user requests

```
guardrails:
  partition_size
    warn_threshold: 1gb
    # when we can support, can add after the fact
    fail_threshold: 2gb
```

If we use the flat name, then the fail also has to be flat (once added), and if we are not consistent this becomes an area for confusion for users: "why is partition_size flat when column_value_size is nested?"

Reply:
@dcapwell  I believe that we should bring that discussion to the dev mailing list because using nested configurations have some impacts on differents patches on which people are currently working (e.g. CASSANDRA-15254, CASSANDRA-15234).
We also need a global agreement for such a change of approach to ensure that everybody is on board and aware of it. Otherwise the yaml configuration might become too confusing.

Reply:
CASSANDRA-15234 is actually where the nested debate started =)

Reply:
pushed a DISCUSS to the dev list